linkerd/stack/src/annotate_error.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{layer, ExtractParam, NewService, Service};
+use std::{
+    error::Error,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// A [`NewService`] that extracts an [`ErrorContext`] implementation from its target,
+/// and produces an [`AnnotateError`] middleware which wraps an inner [`Service`]'s
+/// error type with that [`ErrorContext`].
+pub struct NewAnnotateError<N, C, X = ()> {
+    inner: N,
+    extract: X,
+    _cx: PhantomData<fn(C)>,
+}
+
+/// Provides context to [`Error`]s.
+///
+/// This trait represents a type which can annotate an [`Error`] with
+/// information about the context in which the error occurred, returning a
+/// wrapped [`Self::Error`].
+pub trait ErrorContext {
+    /// The returned error type.
+    type Error: Error + Send + Sync + 'static;
+
+    /// Wraps `error` into a new [`Self::Error`], annotated with context
+    /// provided by `self`.
+    fn annotate<E>(&self, error: E) -> Self::Error
+    where
+        E: Into<linkerd_error::Error>;
+}
+
+/// A [`Service`] middleware which annotates its inner service's error type with
+/// an [`ErrorContext`].
+#[derive(Debug, Clone)]
+pub struct AnnotateError<C, S> {
+    inner: S,
+    context: Arc<C>,
+}
+
+/// Futures returned by [`AnnotateError`].
+#[derive(Debug)]
+#[pin_project::pin_project]
+pub struct ResponseFuture<F, C> {
+    #[pin]
+    f: F,
+    context: Arc<C>,
+}
+
+/// An [`ErrorContext`] implementation that wraps errors in a new error type
+/// that implements `From<(&T, E)>`, where `T` is a stack target.
+pub struct FromTarget<T, E> {
+    target: T,
+    _err: PhantomData<fn(E)>,
+}
+
+type NewFromTarget<N, T, E> = NewAnnotateError<N, FromTarget<T, E>, fn(&T) -> FromTarget<T, E>>;
+
+// === impl NewAnnotateError ===
+
+/// Returns a `Layer` that produces [`NewAnnotateError`] middleware which
+/// annotate errors by constructing an `E`-typed error that implements
+/// `From<(&T, Error)>`, where `T` is a stack target.
+///
+/// `T` must implement [`Clone`]
+pub fn layer_from_target<E, T, N>() -> impl layer::Layer<N, Service = NewFromTarget<N, T, E>> + Clone
```

Comment:
Is there any reason this can't be hung off of NewAnnotateError, too? Seems like it should be doable... I'd expect that to slightly simplify imports for callers.

Reply:
i thought it wouldn't work because the impl has to be generic over `T`, which isn't one of `NewAnnotateError`'s type params, but it actually works fine. I changed it.

Reply:
oh, for the record, i remembered why this was a free function: it's to avoid this: https://github.com/linkerd/linkerd2-proxy/blob/d05b24adb01dce3256a0d69f496fa52dd83b515f/linkerd/app/outbound/src/http/endpoint.rs#L67-L71

ideally, we would want the invocation of `layer_from_target` to be generic over the `Error` type that's being made, rather than the (more complex) type of the _machinery for making that error_, so that we wouldn't have to write out the whole `FromTarget<T, E>` thing and could just write the name of the error type. But when the layer is constructed by an associated function on the `NewService` impl, it's not actually generic over the error type.

not a big deal, just figured i'd mention it.